In [27]:
import re
from test_utilities import run_tests_params

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 2: I Was Told There Would Be No Math ---</h2><p>The elves are running low on wrapping paper, and so they need to submit an order for more.  They have a list of the dimensions (length <code>l</code>, width <code>w</code>, and height <code>h</code>) of each present, and only want to order exactly as much as they need.</p>
<p>Fortunately, every present is a box (a perfect <a href="https://en.wikipedia.org/wiki/Cuboid#Rectangular_cuboid">right rectangular prism</a>), which makes calculating the required wrapping paper for each gift a little easier: find the surface area of the box, which is <code>2*l*w + 2*w*h + 2*h*l</code>.  The elves also need a little extra paper for each present: the area of the smallest side.</p>
<p>For example:</p>
<ul>
<li>A present with dimensions <code>2x3x4</code> requires <code>2*6 + 2*12 + 2*8 = 52</code> square feet of wrapping paper plus <code>6</code> square feet of slack, for a total of <code>58</code> square feet.</li>
<li>A present with dimensions <code>1x1x10</code> requires <code>2*1 + 2*10 + 2*10 = 42</code> square feet of wrapping paper plus <code>1</code> square foot of slack, for a total of <code>43</code> square feet.</li>
</ul>
<p>All numbers in the elves' list are in <span title="Yes, I realize most of these presents are luxury yachts.">feet</span>.  How many total <em>square feet of wrapping paper</em> should they order?</p>
</article>


In [28]:
def packing_paper(l: int, w: int, h: int) -> int:
    return 2 * l * w + 2 * w * h + 2 * h * l + min(l * w, w * h, h * l)


assert packing_paper(2, 3, 4) == 58
assert packing_paper(1, 1, 10) == 43

In [29]:
s = """
2x3x4
1x1x10
"""


def total_square_feet_of_wrapping_paper(s: str) -> int:
    return sum(
        packing_paper(*[int(i) for i in line.split("x")])
        for line in s.strip().splitlines()
    )


assert total_square_feet_of_wrapping_paper(s) == 58 + 43

In [30]:
with open("../input/day2.txt") as f:
    print(total_square_feet_of_wrapping_paper(f.read()))

1586300


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1586300</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The elves are also running low on ribbon.  Ribbon is all the same width, so they only have to worry about the length they need to order, which they would again like to be exact.</p>
<p>The ribbon required to wrap a present is the shortest distance around its sides, or the smallest perimeter of any one face.  Each present also requires a bow made out of ribbon as well; the feet of ribbon required for the perfect bow is equal to the cubic feet of volume of the present.  Don't ask how they tie the bow, though; they'll never tell.</p>
<p>For example:</p>
<ul>
<li>A present with dimensions <code>2x3x4</code> requires <code>2+2+3+3 = 10</code> feet of ribbon to wrap the present plus <code>2*3*4 = 24</code> feet of ribbon for the bow, for a total of <code>34</code> feet.</li>
<li>A present with dimensions <code>1x1x10</code> requires <code>1+1+1+1 = 4</code> feet of ribbon to wrap the present plus <code>1*1*10 = 10</code> feet of ribbon for the bow, for a total of <code>14</code> feet.</li>
</ul>
<p>How many total <em>feet of ribbon</em> should they order?</p>
</article>

</main>


In [31]:
s = """
2x3x4
1x1x10
"""


def ribbon(l: int, w: int, h: int) -> int:
    return 2 * sum(sorted((l, w, h))[:2]) + l * w * h


def total_ribbon(s: str) -> int:
    return sum(
        ribbon(*[int(i) for i in line.split("x")]) for line in s.strip().splitlines()
    )


assert total_ribbon(s) == 34 + 14, str(total_ribbon(s))

In [32]:
with open("../input/day2.txt") as f:
    print(total_ribbon(f.read()))

3737498


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>3737498</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="2/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22I+Was+Told+There+Would+Be+No+Math%22+%2D+Day+2+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F2&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22I+Was+Told+There+Would+Be+No+Math%22+%2D+Day+2+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F2';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
